### PR TITLE
Simulation framework for multi-client server testing

### PR DESCRIPTION
**Pull request by sekulicd**
This introduces a simulation framework designed to test the server by simulating multiple clients performing various actions over several rounds. The framework reads simulation configurations from YAML files, validates them against a predefined schema, and executes the simulation accordingly.

simulation1.yaml is configured as per [this](https://github.com/ark-network/ark/issues/342) requirement, but this was created with idea that any scenario can be configured.

@tiero @altafan @louisinger @bordalix please review.

### CODE REVIEW

File: simulation/main.go [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+	if err := tmpfile.Close(); err != nil {
+		log.Fatal(err)
+	}
+
+	log.Infof("start building ARKD docker container ...")
+	if _, err := utils.RunCommand("docker", "compose", "-f", composePath, "--env-file", tmpfile.Name(), "up", "-d", "--build"); err != nil {
+		log.Fatal(err)
+	}
+
+	time.Sleep(10 * time.Second)
```
Comment (tiero) [inline simulation/main.go]: Do we really need 10 second wait?

File: simulation/schema.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+version: string()
+
+asp:
```
Comment (tiero) [inline simulation/schema.yaml]: Rename to server

File: simulation/main.go [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+		go func(id string, initialFunding float64) {
+			defer wg.Done()
+			cl, err := setupArkClient()
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			usersMtx.Lock()
+			users[id] = User{
+				aspClient:      cl,
```
Comment (tiero) [inline simulation/main.go]: Just client?

File: simulation/main.go [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+		InitialFunding float64 `yaml:"initial_funding,omitempty"`
+	} `yaml:"clients"`
+	Rounds []struct {
+		Number  int                      `yaml:"number"`
+		Actions map[string][]interface{} `yaml:"actions"`
+	} `yaml:"rounds"`
+}
+
+type User struct {
+	aspClient      arksdk.ArkClient
```
Comment (tiero) [inline simulation/main.go]: client

File: server/test/e2e/test_utils.go [hunk 5]
```diff
@@ -104,3 +107,97 @@ func newCommand(name string, arg ...string) *exec.Cmd {
 	cmd := exec.Command(name, arg...)
 	return cmd
 }
+
+func SetupAspWalletCovenantless(initFunding float64) error {
```
Comment (tiero) [inline server/test/e2e/test_utils.go]: SetupServer...

File: simulation/main.go (hunk 1 above)
Comment (sekulicd) [reply simulation/main.go]: Its same setup in existing e2e test, [here](https://github.com/sekulicd/ark/blob/132e9ce208a567589c6660bf8f9b15ddb82b24e6/server/test/e2e/covenantless/e2e_test.go#L35)

File: simulation/schema.yaml (hunk 2 above)
Comment (sekulicd) [reply simulation/schema.yaml]: Done

File: simulation/main.go (hunk 3 above)
Comment (sekulicd) [reply simulation/main.go]: Done

File: simulation/main.go (hunk 4 above)
Comment (sekulicd) [reply simulation/main.go]: Done

File: server/test/e2e/test_utils.go (hunk 5 above)
Comment (sekulicd) [reply server/test/e2e/test_utils.go]: Done

**Comment by sekulicd** [comment]
> I would not bump version in this PR

Reverted

File: simulation/README.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
+# Simulation Framework for Server Testing
+
+This simulation framework is designed to test the **ASP (Ark Server Provider)** by simulating multiple clients performing
```
Comment (tiero) [inline simulation/README.md]:
```suggestion
This simulation framework is designed to test the **Ark Server** by simulating multiple clients performing 
```

File: simulation/main.go [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+	}
+
+	// Store the temporary directory path for later cleanup
+	tempDirsMutex.Lock()
+	tempDirs = append(tempDirs, tempDir)
+	tempDirsMutex.Unlock()
+
+	appDataStore, err := store.NewStore(store.Config{
+		ConfigStoreType:  types.FileStore,
+		AppDataStoreType: types.KVStore,
```
Comment (altafan) [inline simulation/main.go]: we could use inmemory store and and prevent handling temp dirs, no need for appdata store as the clients are initialized without tx feed in client.Init():
```suggestion
		ConfigStoreType:  types.InMemoryStore,
```

File: simulation/schema.yaml [hunk 8]
```diff
@@ -0,0 +1,18 @@
+version: string()
+
+server:
+  network: enum('regtest', 'testnet', 'mainnet')
+  initial_funding: float(min=0)
+  round_interval: int(min=1)
```
Comment (altafan) [inline simulation/schema.yaml:6]:
```suggestion
  round_interval: int(min=5)
```

File: simulation/README.md (hunk 6 above)
Comment (sekulicd) [reply simulation/README.md]: Done

File: simulation/main.go [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	log.Infof("Simulation Version: %s\n", simulation.Version)
+	log.Infof("ASP Network: %s\n", simulation.Server.Network)
+	log.Infof("Number of Clients: %d\n", len(simulation.Clients))
+	log.Infof("Number of Rounds: %d\n", len(simulation.Rounds))
+
+	roundLifetime := fmt.Sprintf("%d", simulation.Server.RoundInterval)
```
Comment (altafan) [inline simulation/main.go]: It seems to me that the content of the docker-env temp file would look like the following with a round interval of let's say 10 seconds:

```
10
```

instead this is what we want?

```
ARK_ROUND_INTERVAL=10
```

```suggestion
	roundInterval := fmt.Sprintf("ARK_ROUND_INTERVAL=%d", simulation.Server.RoundInterval)
```

File: simulation/main.go [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+
+	log.Infoln("ASP wallet initialized")
+
+	go func() {
+		if err := utils.GenerateBlock(); err != nil {
+			log.Fatal(err)
+		}
+
+		time.Sleep(5 * time.Second)
+	}()
```
Comment (altafan) [inline simulation/main.go]: The sleeping time here is pointless as the block is generated in a go routine. Depending on whether we want the operation done in sync, either remove the go routine or the sleeping time.

File: simulation/main.go (hunk 9 above)
Comment (sekulicd) [reply simulation/main.go]: Done

File: simulation/main.go (hunk 7 above)
Comment (sekulicd) [reply simulation/main.go]: Done

File: simulation/main.go (hunk 10 above)
Comment (sekulicd) [reply simulation/main.go]: Check now, i was thinking to gen block every sec, @altafan is this ok?

Comment (altafan) [reply simulation/main.go]: now looks good :)

File: simulation/schema.yaml [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+clients:
+  - id: string()
+    initial_funding: float(min=0, required=false)
+
+rounds:
+  - number: int(min=1)
+    actions:
+      regex('^client_\d+$'):
+        - type: enum('Onboard', 'SendAsync', 'SendOffchain', 'SendOnchain', 'Claim', 'CollaborativeExit', 'UnilateralExit')
```
Comment (bordalix) [inline simulation/schema.yaml]: In order to maintain coherence with method's names, I think "CollaborativeExit" should be renamed to "CollaborativeRedeem"

```suggestion
        - type: enum('Onboard', 'SendAsync', 'SendOffchain', 'SendOnchain', 'Claim', 'CollaborativeRedeem', 'UnilateralExit')
```

**Review by bordalix — APPROVED** [review]

File: simulation/schema.yaml (hunk 11 above)
Comment (tiero) [reply simulation/schema.yaml]: but redeem can be confused with Redeem Tx for oor?

File: docker-compose.regtest.yml [hunk 12]
```diff
@@ -31,9 +31,7 @@ services:
       - ARK_ROUND_INTERVAL=10
       - ARK_NETWORK=liquidregtest
       - ARK_LOG_LEVEL=5
-      - ARK_ESPLORA_URL=http://chopsticks-liquid:3000
-      - ARK_ROUND_LIFETIME=20
-      - ARK_SCHEDULER_TYPE=block
```
Comment (louisinger) [inline docker-compose.regtest.yml:36]: Removing `ARK_ROUND_LIFETIME` & `ARK_SCHEDULER_TYPE` will make the coveanant "TestSweep" test fails

**Comment by tiero** [comment]
Remember to squash the history locally in a single meaningful commit after the work is done.

**Pull request closed without merging**